remove special grad_mode tls handling

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63117
* __->__ #63116

This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

Differential Revision: [D30388098](https://our.internmc.facebook.com/intern/diff/D30388098)

albanD committed:
remove special grad_mode tls handling

[ghstack-poisoned]

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

[ghstack-poisoned]

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

[ghstack-poisoned]

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

[ghstack-poisoned]

soulitzer commented on aten/src/ATen/ThreadLocalState.cpp:
```diff
@@ -8,32 +8,28 @@
       autograd_tls_(c10::AutogradTLS::get_mode()) {
   rf_tls_ = at::get_record_function_tls_();
-
-#if !defined(CAFFE2_IS_XPLAT_BUILD) && !defined(C10_MOBILE)
```
I guess this PR won't affect mobile/xplat builds, but just curious why we had this exception previously.

albanD replied:
My guess is that the mobile build doesn't come with cuda so trying to set the GradMode would raise an error there.

soulitzer reviewed (dismissed):
Nice simplification!

soulitzer commented:
Looks like updates need to be done in the distributed engine side as well

albanD commented:
For some reason ghstack didn't like my rebasing work and pushed a brand new stack (maybe I shouldn't have upgraded ghstack?)
In any case, the new version of this PR is here: https://github.com/pytorch/pytorch/pull/63318

Sorry for the noise.

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

[ghstack-poisoned]

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

[ghstack-poisoned]

soulitzer reviewed (approved):
LGTM

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

[ghstack-poisoned]

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388098).

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

Differential Revision: [D30388098](https://our.internmc.facebook.com/intern/diff/D30388098)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388098).

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

Differential Revision: [D30388098](https://our.internmc.facebook.com/intern/diff/D30388098)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388098).

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

Differential Revision: [D30388098](https://our.internmc.facebook.com/intern/diff/D30388098)

[ghstack-poisoned]

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

Differential Revision: [D30388098](https://our.internmc.facebook.com/intern/diff/D30388098)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388098).

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

Differential Revision: [D30388098](https://our.internmc.facebook.com/intern/diff/D30388098)

[ghstack-poisoned]

albanD committed:
Update on "remove special grad_mode tls handling"


This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

Differential Revision: [D30388098](https://our.internmc.facebook.com/intern/diff/D30388098)

[ghstack-poisoned]
